rhino/src/main/java/org/mozilla/javascript/AbstractEcmaObjectOperations.java

```diff
@@ -232,4 +242,65 @@ static void put(Context cx, Scriptable o, int p, Object v, boolean isThrow) {
             base.put(p, o, v);
         }
     }
+
+    /**
+     * Implement the ECMAScript abstract operation "GroupBy" defined in section 7.3.35 of ECMA262.
+     *
+     * @param cx
+     * @param scope
+     * @param items
+     * @param callback
+     * @param keyCoercion
+     * @see <a href="https://tc39.es/ecma262/#sec-groupby"></a>
+     */
+    static Map<Object, List<Object>> groupBy(
+            Context cx,
+            Scriptable scope,
+            IdFunctionObject f,
+            Object items,
+            Object callback,
+            KEY_COERCION keyCoercion) {
+        if (cx.getLanguageVersion() >= Context.VERSION_ES6) {
+            ScriptRuntimeES6.requireObjectCoercible(cx, items, f);
+        }
+        if (!(callback instanceof Callable)) {
+            throw ScriptRuntime.typeErrorById(
+                    "msg.isnt.function", callback, ScriptRuntime.typeof(callback));
+        }
+
+        // LinkedHashMap used to preserve key creation order
+        Map<Object, List<Object>> groups = new LinkedHashMap<>();
+        final Object iterator = ScriptRuntime.callIterator(items, cx, scope);
+        try (IteratorLikeIterable it = new IteratorLikeIterable(cx, scope, iterator)) {
+            double k = 0;
+            for (Object o : it) {
+                if (k > NativeNumber.MAX_SAFE_INTEGER) {
+                    it.close();
+                    throw ScriptRuntime.typeError("Too many values to iterate");
+                }
+
+                Object[] args = {o, k};
+                Object key =
+                        ((Callable) callback).call(cx, scope, Undefined.SCRIPTABLE_UNDEFINED, args);
+                if (keyCoercion == KEY_COERCION.PROPERTY) {
+                    if (!ScriptRuntime.isSymbol(key)) {
+                        key = ScriptRuntime.toString(key);
+                    }
+                } else {
+                    assert keyCoercion == KEY_COERCION.COLLECTION;
+                    if ((key instanceof Number)
+                            && ((Number) key).doubleValue() == ScriptRuntime.negativeZero) {
+                        key = ScriptRuntime.zeroObj;
+                    }
+                }
+
+                groups.putIfAbsent(key, new ArrayList<>());
+                groups.get(key).add(o);
```

Comment:
You can (and should) do this with only one trip through the hash table -- use "computeIfAbsent", which will return either the new ArrayList or the existing value, and then you can add directly to that, like this (give or take some syntax):

a = groups.computeIfAbsent(key, (k) -> new ArrayList<>());
a.add(o)

That saves you a hash lookup on every iteration which will be significant.